### PR TITLE
Geometry altitude integration

### PR DESCRIPTION
**Pull request by signedav**
## 3D geometry for tables
- reach_point
- wastewater_node
- cover

### Means: 
New geometry type and merging from existing data with Z value taken from `level` attribute - if it's null, the value 'NaN' is written into Z. 

## Synchronization functionality implemented for:
- cover
- wastewater_node
- vw_qgep_reach

### Means: 
On insert  set the Z geometry according the level on update it should update the Z according to the level in case the level changed and the Z did not or it should update the level in case the Z changed and the level did not. When both change, level is dominating.

**cover on INSERT**

take `level` (even when NULL or nothing) as Z

**wastewater_node on INSERT**

take `bottom level` (even when NULL or nothing) as Z

**reach and reach_point on INSERT**

take `level` of `reach_point` (even when NULL or nothing) as Z of `reach_point` and `reach`

**cover ON UPDATE**

if `level` changed, take `level` as Z / if not: if Z changed, take Z as `level`

**wastewater_node on UPDATE**

if `bottom_level` changed, take `bottom_level` as Z / if not: if Z changed, take Z as `bottom_level`

**reach and reach_point on UPDATE**

if `level` of `reach_point` changed, take `level` as Z of `reach_point` and `reach` / if not: if Z changed, take Z as `level` of `reach_point`

on vw_qgep_reach ON INSERT trigger:
- new: update progression_geometry according to rp_from_level und rp_to_level
- existing: rp_from_level und rp_to_level -> reach_point.level
- existing: progression_geometry -> reach_point.situation_geometry

on vw_qgep_reach ON UPDATE rule:
- new: if rp_from_level changed ->  progression_geometry / if not but progression_geometry Z changed -> rp_from_level
- new: if rp_to_level changed ->  progression_geometry / if not but progression_geometry Z changed -> rp_to_level
- existing: rp_from_level und rp_to_level -> reach_point.level
- existing: progression_geometry -> reach_point.situation_geometry

### Testing
Unit tests for the functionality above.

### CODE REVIEW

File: delta/delta_1.1.2_geometry_altitude.post.sql [hunk 1]
```diff
@@ -0,0 +1,1401 @@
...
+
+
+-----------------------------------------------
+-- View: qgep_od.vw_qgep_reach and rules and triggers
+-----------------------------------------------
+
+CREATE OR REPLACE VIEW qgep_od.vw_qgep_reach AS
+
+/* WITH active_maintenance_event AS (
+  SELECT me.obj_id, me.identifier, me.active_zone, mews.fk_wastewater_structure FROM qgep_od.maintenance_event me
```
Comment (lbartoletti) [inline delta/delta_1.1.2_geometry_altitude.post.sql]: Was a test? Can you remove it?

File: delta/delta_1.1.2_geometry_altitude.post.sql [hunk 2]
```diff
@@ -0,0 +1,1401 @@
...
+    rp_to.level AS rp_to_level,
+    rp_to.outlet_shape AS rp_to_outlet_shape,
+    rp_to.position_of_connection AS rp_to_position_of_connection,
+    rp_to.remark AS rp_to_remark,
+    rp_to.last_modification AS rp_to_last_modification,
+    rp_to.fk_dataowner AS rp_to_fk_dataowner,
+    rp_to.fk_provider AS rp_to_fk_provider,
+    rp_to.fk_wastewater_networkelement AS rp_to_fk_wastewater_networkelement
+    /* am.obj_id AS me_obj_id,
+    am.active_zone AS me_active_zone,
```
Comment (lbartoletti) [inline delta/delta_1.1.2_geometry_altitude.post.sql]: same?

File: delta/delta_1.1.2_geometry_altitude.post.sql [hunk 3]
```diff
@@ -0,0 +1,1401 @@
...
+    am.identifier AS me_identifier */
+   FROM qgep_od.reach re
+     LEFT JOIN qgep_od.wastewater_networkelement ne ON ne.obj_id = re.obj_id
+     LEFT JOIN qgep_od.reach_point rp_from ON rp_from.obj_id = re.fk_reach_point_from
+     LEFT JOIN qgep_od.reach_point rp_to ON rp_to.obj_id = re.fk_reach_point_to
+     LEFT JOIN qgep_od.wastewater_structure ws ON ne.fk_wastewater_structure = ws.obj_id
+     LEFT JOIN qgep_od.channel ch ON ch.obj_id = ws.obj_id
+     LEFT JOIN qgep_od.pipe_profile pp ON re.fk_pipe_profile = pp.obj_id;
+     /* LEFT JOIN active_maintenance_event am ON am.fk_wastewater_structure = ch.obj_id; */
```
Comment (lbartoletti) [inline delta/delta_1.1.2_geometry_altitude.post.sql]: same?

File: delta/delta_1.1.2_geometry_altitude.post.sql [hunk 4]
```diff
@@ -0,0 +1,1401 @@
...
+  EXECUTE PROCEDURE qgep_od.vw_qgep_reach_on_update();
+
+-- REACH DELETE
+-- Rule: vw_qgep_reach_on_delete()
+
+CREATE OR REPLACE RULE vw_qgep_reach_on_delete AS ON DELETE TO qgep_od.vw_qgep_reach DO INSTEAD (
+  DELETE FROM qgep_od.reach WHERE obj_id = OLD.obj_id;
+);
+
+--missing: delete also connected wastewater_structure (and subclass channel or other), structure_parts, re_maintenance_events
```
Comment (lbartoletti) [inline delta/delta_1.1.2_geometry_altitude.post.sql]: Is a TODO?

**Comment by lbartoletti** [comment]
cc @ponceta @kandre can you tests too?

**Comment by lbartoletti** [comment]
Also, I opened this issue https://github.com/QGEP/qgepplugin/issues/32

File: delta/delta_1.1.2_geometry_altitude.post.sql (hunk 1 above)
Comment (signedav) [reply delta/delta_1.1.2_geometry_altitude.post.sql]: I can remove it. I took this comment from the original file where this view is created https://github.com/QGEP/datamodel/blob/master/view/vw_qgep_reach.sql#L6 but I agree it does not need to be in the delta as well.

File: delta/delta_1.1.2_geometry_altitude.post.sql (hunk 4 above)
Comment (signedav) [reply delta/delta_1.1.2_geometry_altitude.post.sql]: This is as well a comment I took from the original script. It's a todo I guess, but not one of this request...

**Comment by signedav** [comment]
> Also, I opened this issue [QGEP/qgepplugin#32](https://github.com/QGEP/qgepplugin/issues/32)
This problem appeared only with this implementation?

**Comment by m-kuhn** [comment]
Thanks a lot for the implementation and review.

Wizard related things (like [QGEP/qgepplugin#32](https://github.com/QGEP/qgepplugin/issues/32)) are out of scope for this development.

**Comment by ponceta** [comment]
@signedav Is the update of the Z values in the newly created 3D geometries done using the existing level attributes for existing data in the delta? Or is it a separated SQL query to run afterwards?

I did not find it.

**Comment by signedav** [comment]
Not sure if I understand you correctly @ponceta 
but here https://github.com/QGEP/datamodel/pull/90/files#diff-3bbd562083345cd31950a05728dfc7ceR11
it takes the level if not NULL as the Z value from existing data:

**Comment by ponceta** [comment]
@signedav that's what I meant, indeed, thanks!

Looks good to me!

**Comment by m-kuhn** [comment]
Thanks everyone!!

**Comment by signedav** [comment]
Thanks guys 🙂

**Pull request merged**